feat: add kafka backfill frontend

xxchan opened the pull request:
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

<!--

**Please do not leave this empty!**

Please explain **IN DETAIL** what the changes are in this PR and why they are needed:

- Summarize your change (**mandatory**)
- How does this PR work? Need a brief introduction for the changed logic (optional)
- Describe clearly one logical change and avoid lazy messages (optional)
- Describe any limitations of the current code (optional)
- Refer to a related PR or issue link (optional)

-->

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)
<!-- To manually trigger a benchmark, please check out [Notion](https://www.notion.so/risingwave-labs/Manually-trigger-nexmark-performance-dashboard-test-b784f1eae1cf48889b2645d020b6b7d3). -->
- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.


<!--
Please create a release note for your changes.

Discuss technical details in the "What's changed" section, and
focus on the impact on users in the release note.

You should also mention the environment or conditions where the impact may occur.
-->

</details>

xxchan commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/15602?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#15602** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/15602?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* **#15635** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/15635?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#15632** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/15632?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @xxchan and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

xxchan commented on src/stream/src/executor/source/source_backfill_executor.rs:
```diff
@@ -611,11 +611,11 @@ impl<S: StateStore> SourceBackfillExecutorInner<S> {
-                            .state_store
-                            .commit(barrier.epoch)
-                            .await?;
                     }
+                    self.backfill_state_store
+                        .state_store
+                        .commit(barrier.epoch)
+                        .await?;
```
It seems commit must make sure previous epoch is also committed. So made this change

fuyufjh commented on src/connector/src/parser/additional_columns.rs:
```diff
@@ -185,6 +185,67 @@ pub fn build_additional_column_catalog(
+            })
+            .collect()
+    };
+    assert_eq!(additional_columns.len(), 2);
```
Looks like the order between `partition/file` and `offset` is implicitly assumed. How about adding an assertion such as:


```
    assert_eq!(additional_columns[0] is partition or file); // pseudocode
    assert_eq!(additional_columns[1] is offset);
```

fuyufjh commented on src/frontend/src/optimizer/plan_node/logical_source_backfill.rs:
```diff
@@ -0,0 +1,207 @@
+        let mut column_catalog = source_catalog.columns.clone();
+        let row_id_index = source_catalog.row_id_index;
+
+        let kafka_timestamp_range = (Bound::Unbounded, Bound::Unbounded);
```
Hard-coded?

xxchan replied:
This is copy-pasted code from LogicalSource. And the kafka timestamp stuff does need refinement. https://github.com/risingwavelabs/risingwave/issues/14636 https://github.com/risingwavelabs/risingwave/issues/14627

fuyufjh commented on src/frontend/src/optimizer/plan_node/stream_source_backfill.rs:
```diff
@@ -0,0 +1,186 @@
+                    .to_internal_table_prost(),
+            ),
+            info: Some(source_catalog.info.clone()),
+            // XXX: what's the usage of this?
```
You mean `row_id`? By design, rows from source should be assigned unique row_ids as its PK.

fuyufjh commented on src/frontend/src/optimizer/plan_node/logical_source_backfill.rs:
```diff
@@ -0,0 +1,207 @@
+use crate::utils::{ColIndexMapping, Condition};
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct LogicalSourceBackfill {
```
IIUC, `LogicalSourceBackfill` will be converted to both `Backfill` and `Upstream`, so it actually should be named `LogicalSourceScan`?

fuyufjh commented on src/frontend/src/optimizer/plan_node/logical_source.rs:
How to cover it in test? 🤔 planner test or e2e?

fuyufjh commented on proto/catalog.proto:
```diff
@@ -62,9 +62,16 @@ message StreamSourceInfo {
+  // Only used when `has_streaming_job` is `true`.
+  // If `false`, `requires_singleton` will be set in the stream plan.
+  bool is_distributed = 15;
+  reserved "cdc_source_job"; // deprecated
```
Deprecating `cdc_source_job` and adding `has_streaming_job` will break backward compability of existing CDC jobs, right?

Why not renaming `cdc_source_job` to `has_streaming_job`?

xxchan replied:
> Why not renaming cdc_source_job to has_streaming_job?

This is indeed the change I made. Number `13` is renamed. 

> Changing the field name could be breaking under SQL meta backend.

Indeed. 😕 Need to come up with another plan.

tabVersion replied:
do you already have one? Keeping it or introducing another one dedicated to this case both LGTM

xxchan replied:
New plan:
- keep `cdc_source_job`. New & old CDC sources will both set this field. This field is only used to set `require_singleton`
- Change `is_shared` to `optional` (explicit presence). Old CDC sources will have this field set to `None`. New CDC sources will set to `Some(true)`.
  - added method `is_shared_compatibile` to test `cdc_source_job` first. TBH, this is a little awkward. Better ideas are welcome.
  - Strictly speaking, it's not necessary to use `optional`, `(cdc_source_job, is_shared) = (true,false)` can also work. But I feel this can make it (just slightly) less error-prone.

---
Old plan:
- Rename `cdc_source_job` to `is_shared`. This is more elegant, but breaks JSON compatibility for SQL meta backend.
- Added `is_distributed` to set `require_singleton` in fragmenter for CDC source.

fuyufjh replied:
> Rename `cdc_source_job` to `is_shared`. This is more elegant, but breaks JSON compatibility for SQL meta backend.


Maybe off-topic... I think storing protobuf into JSON sounds bad.

<img width="1010" alt="image" src="https://github.com/risingwavelabs/risingwave/assets/10192522/336121d2-c040-4493-9ad0-dbfdd5abba48">

https://protobuf.dev/programming-guides/dos-donts/#text-format-interchange

fuyufjh commented on src/meta/src/stream/source_manager.rs:
```diff
@@ -778,6 +873,63 @@ impl SourceManager {
         Ok(assigned)
     }
 
+    pub async fn allocate_splits_for_backfill(
```
Does it need `migrate_splits_for_backfill` as well? May need to test upon split repartition.

fuyufjh replied:
BTW, please add some rustdocs ❤️

xxchan replied:
> Does it need migrate_splits_for_backfill as well? 

Yes. Reschedule is not fully considered yet. 

> May need to test upon split repartition.

Indeed. BTW, split change is handled by `reassign_splits`. Currently we have no tests for that at all. 😕 Will try to add some.

fuyufjh reviewed (commented):
Looks good

yezizp2012 commented on src/meta/src/manager/metadata.rs:
```diff
@@ -613,6 +613,22 @@ impl MetadataManager {
+                    .await
+            }
+            MetadataManager::V2(_mgr) => {
+                todo!()
```
Can we add the implementation as well in catalog controller? We can simply query the `actor_id` and `upstream_actor_ids` columns of the `actor` table, filter by `fragment_id` and `status`.

yezizp2012 commented on src/meta/src/rpc/ddl_controller.rs:
```diff
@@ -1360,8 +1361,12 @@ impl DdlController {
             .await?;
 
-        let upstream_actors: HashMap<_, _> = upstream_root_fragments
+        // XXX: do we need to filter here?
```
Nope, IIUC we don't need this filter. It could be shared source streaming job. We'd better change the naming in `CreateStreamingJobContext` and `Command::CreateStreamingJob` accordingly.

yezizp2012 commented on src/meta/src/stream/stream_manager.rs:
```diff
@@ -456,7 +456,7 @@ impl GlobalStreamManager {
 
             let dummy_table_id = table_fragments.table_id();
-
+            // TODO: does this need change? for replace_table
```
Yes, the new table has different actor ids, It did not allocate according to the previous one-to-one relationship, but directly called for reallocation. Previously the new replaced table could have a different parallelism from old one, which has been fixed recently. It's acceptable to keep this logic unchanged.

xxchan replied:
Previously I don't know what is "replace table". I found it's for `ALTER TABLE` and sink into table. So it seems no Backfill executor (and also their upstream Source executor) will be involved. 🤔

xxchan replied:
Wait, I'm thinking that even for altering table with connector, it seems the Source actors do not need change? Why do we need to allocate splits……

BugenZhao replied:
All actors in the table job will be rebuilt including the source executors. I guess we have to retrack the split allocation in the source manager as they're assigned a different set of actor ids.

xxchan replied:
Sure. So I think
- `CREATE TABLE` with connector need retrack
- `CREATE SOURCE` (shared) & MV on sources do not need change.

So no changes need to be made in this PR. Do I understand it correctly?

---

> All actors in the table job will be rebuilt including the source executors. 

I don't fully understand the reason behind this though. Is it feasible not to rebuild the source actor? 🤪 Is it because this makes the implementation easier?

yezizp2012 replied:
> 🤪 Is it because this makes the implementation easier?
I'm afraid so. 🥵

tabVersion commented:
Had an offline discussion with @xxchan, if we enable the useable source, there will be real actors running and keep fetching data even if no subsquent MVs. 

Should have an optim to pause the source in this case to save resources, it is ok to impl it in next pr.

yezizp2012 reviewed (commented):
Rest of meta part LGTM. Can we run single node test in this PR? I'm not sure if it could fail it.

BugenZhao commented on src/frontend/src/stream_fragmenter/mod.rs:
```diff
@@ -294,6 +294,7 @@ fn build_fragment(
         }
 
         NodeBody::StreamCdcScan(_) => {
+            // XXX: Should we use a different flag for CDC scan?
```
Depending on whether we find the need to distinguish between them.

xxchan replied:
I think we do need to distinguish them. See the awkward code in `tracking_progress_actor_ids`. But I'm worried about compatibility so don't want to change it.

BugenZhao commented on src/meta/src/stream/stream_graph/actor.rs:
```diff
@@ -254,6 +254,44 @@ impl ActorBuilder {
             }
 
+            // "Leaf" node `SourceBackfill`.
+            NodeBody::SourceBackfill(source_backfill) => {
```
Is there any common snippet to be extracted with `CdcFilter`?

xxchan commented on proto/catalog.proto:
```diff
@@ -62,9 +62,16 @@ message StreamSourceInfo {
+  // Currently, the following sources have streaming jobs:
+  // - Direct CDC sources (mysql & postgresql)
+  // - MQ sources (Kafka, Pulsar, Kinesis, etc.)
+  bool has_streaming_job = 13;
```
Indeed. "shared source" sounds nicer than "reuseable source" 😄

xxchan commented on src/connector/src/parser/additional_columns.rs:
```diff
@@ -185,6 +185,67 @@ pub fn build_additional_column_catalog(
+        let compat_col_types = COMPATIBLE_ADDITIONAL_COLUMNS
+            .get(connector_name)
+            .unwrap_or(&COMMON_COMPATIBLE_ADDITIONAL_COLUMNS);
+        ["partition", "file", "offset"]
```
These code are just moved from `src/connector/src/source/reader/desc.rs` 🙂

xxchan commented on src/frontend/src/optimizer/plan_node/logical_source_backfill.rs:
The plan is different. But it's interesting to consider whether we can merge them.

StrikeW commented on src/connector/src/source/kafka/private_link.rs:
```diff
@@ -70,9 +70,9 @@ impl BrokerAddrRewriter {
-        tracing::info!("[{}] rewrite map {:?}", role, broker_rewrite_map);
         let rewrite_map: ConnectorResult<BTreeMap<BrokerAddr, BrokerAddr>> = broker_rewrite_map
             .map_or(Ok(BTreeMap::new()), |addr_map| {
+                tracing::info!("[{}] rewrite map {:?}", role, addr_map);
```
Why move this log? If the mapping in none, this log won't show up.

xxchan replied:
Why do we need the log when the mapping is none (when private link is not used)? The log is a little annoying (too many)

kwannoel commented on Makefile.toml:
```diff
@@ -86,9 +86,9 @@ if ${is_not_ci}
 
   if ${no_rust_log}
-    set_env RUST_LOG "pgwire_query_log=info"
+    set_env RUST_LOG "pgwire_query_log=info,hyper::client::connect::http=info"
```
Why do we enable `hyper::client::connect::http=info`?

kwannoel commented on e2e_test/source/README.md:
```diff
@@ -0,0 +1,7 @@
+Test in this directory needs some prior setup.
+
+See also `ci/scripts/e2e-source-test.sh`, and `scripts/source`
+
```
I think we can provide the steps to run e2e source test here.

xxchan replied:
I haven't yet tried to find an elegant way to run the tests locally. I'm not sure the script can be used locally in current shape.

kwannoel commented on e2e_test/source/basic/kafka_shared_source.slt:
```diff
@@ -0,0 +1,58 @@
+3 333
+4 4444
+
+# TODO: add more data to the topic and re-check the data. Currently there's no good test infra to do this...
```
Yeah I encounter the same issue, and run my backfill tests separately from e2e tests in `run-backfill-tests.sh`

kwannoel commented on src/common/src/lib.rs:
```diff
@@ -43,6 +43,8 @@
 #![feature(negative_impls)]
 #![feature(bound_map)]
 #![feature(array_methods)]
+#![feature(register_tool)]
```
https://doc.rust-lang.org/beta/unstable-book/language-features/register-tool.html doesn't seem well documented. Could you elaborate more on the usage here?

kwannoel replied:
Out of scope nevermind. It's introduced here: https://github.com/risingwavelabs/risingwave/pull/13763/files#diff-aef7438a31f20c6a6fc81e4f413e8f2231e2fb470823dc18110620ecc0da1843

kwannoel commented on src/common/src/session_config/mod.rs:
```diff
@@ -243,6 +243,13 @@ pub struct ConfigMap {
+    /// When enabled, `CREATE SOURCE` will create a source streaming job, and `CREATE MATERIALIZED VIEWS` from the source
+    /// will forward the data from the same source streaming job, and also backfill prior data from the external source.
+    #[parameter(default = false)]
+    rw_enable_shared_source: bool,
```
PR description needs to be updated, it is called `reusable` there.

xxchan commented:
Changes since last review (It should be OK to review commits since last review)
- Do not add new logical node. new stream node is enough. Added planner tests
- Added basic e2e tests
- Changed catalog handling (do not deprecate `cdc_source_job`)  https://github.com/risingwavelabs/risingwave/pull/15602#discussion_r1545878574

yezizp2012 reviewed (commented):
Meta part LGTM!

fuyufjh commented on proto/catalog.proto:
```diff
@@ -63,8 +63,17 @@ message StreamSourceInfo {
+  // - Direct CDC sources (mysql & postgresql). Backwards compat note: For old CDC job it's `None`; For new CDC job it's `Some(true)`.
+  // - MQ sources (Kafka, Pulsar, Kinesis, etc.)
+  //
+  // **Should also test `cdc_source_job` for backwards compatibility. Use `is_shared_compatible()` instead of this field directly.**
```
Looks error-prone to me. I would rather just reuse `cdc_source_job` even without renaming.

xxchan replied:
😄 Sure that's acceptable to me

fuyufjh replied:
Had better reuse **and** rename 😄

xxchan commented on src/frontend/src/handler/create_source.rs:
```diff
@@ -1313,12 +1313,20 @@ pub async fn handle_create_source(
     let sql_pk_names = bind_sql_pk_names(&stmt.columns, &stmt.constraints)?;
 
     let create_cdc_source_job = with_properties.is_backfillable_cdc_connector();
```
I guess I reverted it by mistake when resolving conflicts 🤡
